shellai/config.py

```diff
@@ -0,0 +1,150 @@
+import json
+import logging
+import sys
+from collections import namedtuple
+from pathlib import Path
+
+if sys.version_info >= (3, 11):
+    import tomllib
+else:
+    import tomli as tomllib
+
+CONFIG_DEFAULT_PATH: Path = Path("~/.config/shellai/config.toml")
+
+# tomllib does not support writting files, so we will create our own.
+CONFIG_TEMPLATE = """
+[output]
+# otherwise recording via script session will be enforced
+enforce_script = {enforce_script}
+# file with output(s) of regular commands (e.g. ls, echo, etc.)
+file = "{output_file}"
+# Keep non-empty if your file contains only output of commands (not prompt itself)
+prompt_separator = "{prompt_separator}"
+
+[history]
+enabled = {enabled}
+file = "{history_file}"
+# max number of queries in history (including responses)
+max_size = {max_size}
+
+[backend]
+endpoint = "{endpoint}"
+
+"""
+
+
+class Output(namedtuple("Output", ["enforce_script", "file", "prompt_separator"])):
+    """This class represents the [output] section of our config.toml file."""
+
+    __slots__ = ()
+
+    def __new__(
```

Comment:
`__slots__` and `__new__` are both non-standard things to use in general Python code.  I think I understand why you're using both of them.  Tell me if I'm right:
* `__slots__` You want to strictly enforce that people cannot add extra attributes to one of these classes at runtime.  (This isn't really Pythonic [The pythonic principle is known as Consenting Adults] but I think it has some justification here.  You're using these `NamedTuple`s as a schema for the config file.  So locking down extra fields could be used to guard against extraneous fields in the config files.
* `__new__` You want to have default values for the fields.  In Python-3.7, this is possible via a parameter to `NamedTuple` but we're targetting Python 3.6 as our lower limit.  So you need a way to emulate that functionality.

If these are both right, I would add comments that explain why we're using these features (and note that we can switch away from using `__new__` when we drop Python 3.6 support.  Usually I'd do that in a base class but you don't have a base class here because you are just using namedtuple so I'm not sure where the best place is.

Reply:
will add the comment